### PR TITLE
[rocksdb] prorotype wrapper for transactional and regular db

### PR DESCRIPTION
**Pull request by phoenix-antigravity**
**context**: in rust rocksdb binding each database is represented by a separate struct. 
And although majority of methods share the same semantic, DBWithThreadMode, OptimisticTransactionDB and TransactionDB are different structs. 
Majority of db structs in the codebase, such as DBMap, have embedded reference to DBWithThreadMode instance.

In order to enable optimistic transactions for some use cases, while reusing existing data structures(e.g. DBMap), PR introduces thin wrapper around various rocksdb structs that unifies the interface.

To recap, for now PR only introduces new unified interface for DBWithThreadMode/OptimisticTransactionDB, switches existing call sites to use new enum and adds a unit test as a reference example of transaction usage

### CODE REVIEW

File: crates/typed-store/src/rocks/mod.rs [hunk 1]
```diff
@@ -92,13 +92,144 @@ macro_rules! reopen {
...
+    pub fn try_catch_up_with_primary(&self) -> Result<(), rocksdb::Error> {
+        delegate_call!(self.try_catch_up_with_primary())
+    }
+
+    pub fn write(&self, batch: rocksdb::WriteBatch) -> Result<(), TypedStoreError> {
+        match self {
+            RocksDB::DBWithThreadMode(db) => Ok(db.write(batch)?),
+            RocksDB::OptimisticTransactionDB(_) => Err(TypedStoreError::RocksDBError(
+                "operation not supported".to_string(),
+            )),
```
Comment (sadhansood) [inline crates/typed-store/src/rocks/mod.rs:198]: We will most definitely need write with a write_batch to work for OptimisticTxDB. It may have its own `WriteBatch` implementation that we need to put under an enum

File: crates/typed-store/src/rocks/mod.rs [hunk 2]
```diff
@@ -92,13 +92,144 @@ macro_rules! reopen {
...
+
+    pub fn raw_iterator_cf<'a: 'b, 'b>(
+        &'a self,
+        cf_handle: &impl AsColumnFamilyRef,
+    ) -> rocksdb::DBRawIteratorWithThreadMode<'b, DBWithThreadMode<MultiThreaded>> {
+        match self {
+            Self::DBWithThreadMode(db) => db.raw_iterator_cf(cf_handle),
+            Self::OptimisticTransactionDB(_) => {
+                panic!("iterator is not implemented for transactional db")
+            }
```
Comment (sadhansood) [inline crates/typed-store/src/rocks/mod.rs:221]: We also need to have a working iter implementation for optimistic tx db

File: crates/typed-store/src/rocks/mod.rs [hunk 3]
```diff
@@ -373,34 +497,41 @@ impl<K, V> DBMap<K, V> {
...
-            .rocksdb_table_readers_usage
-            .with_label_values(&[&db_name])
-            .set(
-                mem_usage_stats
-                    .as_ref()
-                    .map(|x| x.mem_table_readers_total as i64)
-                    .unwrap_or(METRICS_ERROR),
-            );
+
+        if let RocksDB::DBWithThreadMode(ref rocksdb) = **rocksdb {
```
Comment (sadhansood) [inline crates/typed-store/src/rocks/mod.rs:501]: Are we sure these metrics are not available for optimistc tx db?

File: crates/typed-store/src/rocks/mod.rs [hunk 4]
```diff
@@ -456,7 +587,7 @@ impl<K, V> DBMap<K, V> {
 /// ```
 ///
 pub struct DBBatch {
-    rocksdb: Arc<rocksdb::DBWithThreadMode<MultiThreaded>>,
+    rocksdb: Arc<RocksDB>,
     batch: WriteBatch,
```
Comment (sadhansood) [inline crates/typed-store/src/rocks/mod.rs:591]: I think this `WriteBatch` will also need to be an enum - for different db types

File: crates/typed-store/src/rocks/mod.rs [hunk 5]
```diff
@@ -578,6 +709,45 @@ impl DBBatch {
...
+
+impl<'a> DBTransaction<'a> {
+    pub fn new(db: &'a Arc<RocksDB>) -> Result<Self, TypedStoreError> {
+        Ok(Self {
+            rocksdb: db.clone(),
+            transaction: db.transaction()?,
+        })
+    }
+
+    pub fn insert_batch<J: Borrow<K>, K: Serialize, U: Borrow<V>, V: Serialize>(
```
Comment (sadhansood) [inline crates/typed-store/src/rocks/mod.rs:725]: We will also need a `delete_batch` and `delete_range` for txns

File: crates/typed-store/src/rocks/mod.rs [hunk 6]
```diff
@@ -578,6 +709,45 @@ impl DBBatch {
...
+                let k_buf = be_fix_int_ser(k.borrow())?;
+                let v_buf = bincode::serialize(v.borrow())?;
+                self.transaction.put_cf(&db.cf(), k_buf, v_buf)?;
+                Ok(())
+            })?;
+        Ok(self)
+    }
+
+    pub fn commit(self) -> Result<(), TypedStoreError> {
+        self.transaction.commit()?;
```
Comment (sadhansood) [inline crates/typed-store/src/rocks/mod.rs:746]: We also need to have `get_for_update` and `snapshot` for optimistic transactions. I think rust_rocksdb has these iirc

File: crates/typed-store/src/rocks/mod.rs (hunk 1 above)
Comment (phoenix-antigravity) [reply crates/typed-store/src/rocks/mod.rs:198]: sure, but on high level, I wanted to keep this PR reasonably small. 
Initially I was even thinking about just introducing enum and switching existing use cases to new enum.
So I wanted to ship this with MVP functionality and create a followup with some missing features.
Although some of them I want to discuss(understand better) cause it's not that trivial to add

File: crates/typed-store/src/rocks/mod.rs (hunk 3 above)
Comment (phoenix-antigravity) [reply crates/typed-store/src/rocks/mod.rs:501]: we use external library that only supports RocksDB::DBWithThreadMode. We can put TODO to find another way to collect stats for optimistic db

File: crates/typed-store/src/rocks/mod.rs (hunk 5 above)
Comment (phoenix-antigravity) [reply crates/typed-store/src/rocks/mod.rs:725]: `delete_batch`  for sure, but I was thinking to split PR at this point
`delete_range` is not supported for optimistic transaction db

File: crates/typed-store/src/rocks/mod.rs [hunk 7]
```diff
@@ -578,6 +709,83 @@ impl DBBatch {
...
+            .into_iter()
+            .try_for_each::<_, Result<_, TypedStoreError>>(|k| {
+                let k_buf = be_fix_int_ser(k.borrow())?;
+                self.transaction.delete_cf(&db.cf(), k_buf)?;
+                Ok(())
+            })?;
+        Ok(self)
+    }
+
+    pub fn snapshot(&self) {
```
Comment (sadhansood) [inline crates/typed-store/src/rocks/mod.rs:764]: We don;t need to mention the return type for this function? It's probably not even returning a snapshot object i think

**Review by sadhansood — APPROVED** [review]
Looks good to land once we fix the return type of `snapshot` function. Thanks for your efforts! We should probably add some tests which demonstrate the use of `snapshot` and `get_for_update` but we can do that in follow up PRs (I can help too)
